src/storage/cache/rpc/server.rs

```diff
@@ -0,0 +1,486 @@
+        req_buffer: &[u8],
+        done_tx: mpsc::Sender<Vec<u8>>,
+    );
+}
+
+/// The connection for the RPC server.
+#[derive(Clone, Debug)]
+pub struct RpcServerConnection<T>
+where
+    T: RpcServerConnectionHandler + Send + Sync + 'static,
+{
+    /// The inner connection for the RPC server.
+    inner: Arc<RpcServerConnectionInner<T>>,
+}
+
+///
+#[derive(Debug)]
+pub struct RpcServerConnectionInner<T>
+where
+    T: RpcServerConnectionHandler + Send + Sync + 'static,
+{
+    /// The TCP stream for the connection.
+    stream: UnsafeCell<net::TcpStream>,
+    /// The worker pool for the connection.
+    /// TODO:
+    #[allow(dead_code)]
+    worker_pool: Arc<WorkerPool>,
+    /// Options for the timeout of the connection
+    timeout_options: ServerTimeoutOptions,
+    /// The handler for the connection
+    dispatch_handler: T,
+    /// Response buffer, try to reuse same buffer to reduce memory allocation
+    /// In case of the response is too large, we need to consider the buffer size
+    /// Init size is 4MB
+    /// Besides, we want to reduce memory copy, and read/write the response to the same data buffer
+    req_buf: UnsafeCell<BytesMut>,
+}
+
+/// Current implementation is safe because the stream is only accessed by one thread
+unsafe impl<T> Send for RpcServerConnectionInner<T> where
+    T: RpcServerConnectionHandler + Send + Sync + 'static
+{
+}
+
+/// Current implementation is safe because the stream is only accessed by one thread
+unsafe impl<T> Sync for RpcServerConnectionInner<T> where
+    T: RpcServerConnectionHandler + Send + Sync + 'static
+{
+}
+
+impl<T> RpcServerConnectionInner<T>
+where
+    T: RpcServerConnectionHandler + Send + Sync + 'static,
+{
+    /// Create a new RPC server connection.
+    pub fn new(
+        stream: net::TcpStream,
+        worker_pool: Arc<WorkerPool>,
+        timeout_options: ServerTimeoutOptions,
+        dispatch_handler: T,
+    ) -> Self {
+        Self {
+            stream: UnsafeCell::new(stream),
+            worker_pool,
+            timeout_options,
+            dispatch_handler,
+            req_buf: UnsafeCell::new(BytesMut::with_capacity(
+                common::DEFAULT_TCP_REQUEST_BUFFER_SIZE,
+            )),
+        }
+    }
+
+    /// Recv request header from the stream
+    pub async fn recv_header(&self) -> Result<ReqHeader, RpcError> {
+        // Try to read to buffer
+        match self.recv_len(REQ_HEADER_SIZE).await {
+            Ok(()) => {}
+            Err(err) => {
+                debug!("Failed to receive request header: {:?}", err);
+                return Err(err);
+            }
+        }
+
+        let buffer: &mut BytesMut = unsafe { &mut *self.req_buf.get() };
+        let req_header = ReqHeader::decode(buffer)?;
+        debug!("Received request header: {:?}", req_header);
+
+        Ok(req_header)
+    }
+
+    /// Recv request body from the stream
+    pub async fn recv_len(&self, len: u64) -> Result<(), RpcError> {
```

Comment:
Should add an timeout arg to recv_len().
use idle_timeout for header ,  and use read_timeout for body

Reply:
OK, that make sense for me.